[MISC] Fixing Travis errors with Remark (#320)
bids-standard/bids-specification: CONTRIBUTING.md

GitHub has a [nice introduction](https://help.github.com/articles/github-flow/) to the pull request workflow, but please [get in touch](#get-in-touch) if you have any questions.

## Example pull request
<img align="center" src="https://i.imgur.com/s8yELfK.png" alt="Example-Contribution" width="800"/>  

## Fixing Travis Remark errors

We use a linter called [Remarkjs](https://github.com/remarkjs/remark-lint) to ensure all of our specification documentation is consistent.
We have found this causes typical errors that are flagged by our continous integration service called [Travis CI](https://travis-ci.org/).
This behavior is commonly seen when a PR is opened and Travis returns an error. 
To fix this error please follow this process:

#### 1. Download Remark-CLI and our style guide

> Remark-CLI can be downloaded with the help of [npm](https://www.npmjs.com/). 

Reviewer: ```suggestion
Remark-CLI can be installed via [npm](https://www.npmjs.com/), which is part of
the NodeJS distribution.
```

Once npm has been downloaded, please follow the [installation instructions on Remark's GitHub.](https://github.com/remarkjs/remark/tree/master/packages/remark-cli#install)

> We use a particular style guide for formatting. 

Reviewer: ```suggestion
To install the packages we use for our style guide:
```

> To get this style guide please run: 

Reviewer: ```suggestion
```

```
npm install remark-cli@5.0.0 remark-lint@6.0.2 remark-preset-lint-recommended@3.0.2 remark-preset-lint-markdown-style-guide@2.1.2

> ```

Author: perhaps packaging these up? a potential of this would be needing to maintain that package too. This was grabbed from our https://github.com/bids-standard/bids-specification/blob/master/.travis.yml config line 8. That is the place I know we check our formatting

> ```

Reviewer: I just meant something equivalent to a `requirements.txt` in Python. Here's something that could work:

`npm-requirements.txt`:
```
remark-cli@5.0.0
remark-lint@6.0.2
remark-preset-lint-recommended@3.0.2
remark-preset-lint-markdown-style-guide@2.1.2
```

Then it could be installed locally or on Travis with:

```Shell
cat npm-requirements.txt | xargs npm install
```

But there might be something equivalent in the NodeJS world. A very cursory glance didn't find it.

> ```

Author: yeah we can do that - I think definitely locally and can be extended to Travis. Will ease the process. have added this

> ```

Reviewer 2: I also looked for a `pip install -r requirements.txt` equivalent in the `npm` world and it seems like it doesn't exist. 

>  Is this the only place these versions are recorded, or do we need to keep this in sync with some other document

needs to stay in sync with CircleCI [config.yml](https://github.com/bids-standard/bids-specification/blob/master/.circleci/config.yml), so I suggest that in the CircleCI config, we also use the newly created `npm-requirements.txt` file from this PR. Can you do that @franklin-feingold ?

Also, perhaps we can add a sentence for Windows users, who don't have command line tools like `cat`, `|`, or `xargs` by default: These people should open the `npm-requirements.txt` file and install each line by hand, as a practical solution

> ```

Reviewer: At least with PowerShell, people can do:

```Powershell
npm install @(cat npm-requirements.txt)
```

In bourne-like shells, we could also have people do:

```Bash
npm install $(cat npm-requirements.txt)
```

There's probably some unholy utterance for `cmd.exe`, but I doubt many people using `npm` on Windows won't have access to Bash or Powershell.

> ```

Author: oh yeah thanks I do check remark with circle - added in npm-requirements to circle and the Windows cases Chris identified

> #### 2. Fix the flagged document

Reviewer: ```suggestion
#### 3. Fix the flagged document
```

Please go to the directory where the flagged file is and run remark like this:

```
remark flagged_file.md -o flagged_file_fixed.md
```

Please confirm this has fixed the file. 
To do this, please run this:

```
remark flagged_file_fixed.md --frail
```

> This command will return whether this file can pass Travis.

Reviewer: ```suggestion
This command will indicate whether this file now conforms to the style guide.
```

> If it passes, the `flagged_file_fixed.md` can update the `flagged_file.md`.

Reviewer: Suggesting a code block doesn't work well, so just using straight markdown below:

----

If it passes, replace `flagged_file.md` with the contents of `flagged_file_fixed.md`,
`add` and `commit` the change:

```Shell
mv flagged_file_fixed.md flagged_file.md
git add flagged_file.md
git commit -m 'STY: Fixed Markdown style'
```

> Please ensure the names are the same before pushing the fixed file to your forked repository.

Reviewer: Is this a worry?

Author: shouldn't the files be named the same so it gets updated properly? we don't want the bad and fixed next to each other

Reviewer: Right, it just seems a bit obvious. Whatever, though.

> Once the fixed file has been committed, Travis will run again and should pass.

Reviewer: ```suggestion
Once the fixed file has been pushed to your PR branch, Travis will run again and should pass.
```
